src/fold.jl

```diff
@@ -0,0 +1,320 @@
+
+
+"""
+    minimum(d::Folded)
+
+Returns the lowest value of the domain of the distribution
+"""
+minimum(d::Folded) = minimum(d.included)
+
+
+
+"""
+    maximum(d::Folded)
+
+Returns the highest value of the domain of the distribution
+"""
+maximum(d::Folded) = maximum(d.included)
+
+
+"""
+    rand(::AbstractRNG, d::Folded)
+
+Returns samples from the folded distribution
+"""
+rand(rng::AbstractRNG, d::Folded) = fold_value(rand(rng,d.original), d)
+
+"""
+   logpdf(d::Folded, x::Real)
+
+Returns the logarithm of the pdf of the folded distribution. 
+Returns -Inf if the value is out of the distribution.
+
+Example:
+```julia
+logpdf(folded(Normal(μ,σ),0), -3) # -> -Inf
+logpdf(folded(Normal(1.0,1.0),0), 3) # -> 0.05412479673895295
+``` 
+"""
+function logpdf(d::Folded, x::Real)
+    !insupport(d,x) && return -Inf
+    return logaddexp(logpdf(d.original,x),logpdf(d.original,unfold_value(x,d)))
+end
+
+
+"""
+   pdf(d::Folded, x::Real)
+
+Returns the the pdf of the folded distribution. This is simply the
+sum of the value of the pdf of the original distribution at two locations:
+one at `x` and the other at the reflected point `2c-x`, where `c` is the crease.
+
+Example:
+```julia
+pdf(folded(Normal(μ,σ),0), -3) # -> 0.0
+pdf(folded(Normal(1.0,1.0),0), 3) # -> 0.05412479673895295
+pdf(folded(Normal(3,4),0.0),0.1) ≈ pdf(Normal(3,4),-0.1) + pdf(Normal(3,4),0.1) # -> true
+```
+"""
+function pdf(d::Folded, x::Real)
+    !insupport(d,x) && return zero(x)
+    #return pdf(d.original,x) + pdf(d.original,unfold_value(x,d)) -- removed for numerical stability
+    return exp(logpdf(d,x))
+end
+
+"""
+    cdf(d::Folded, x::Real) 
+
+Example:
+cdf(folded(Beta(α,β),0.5), -3) # -> 0.0
+cdf(folded(Beta(α,β),0.5), 3) # -> 1.0
+cdf(folded(Normal(1.0,1.0),0), 1.0) # -> 0.4772498680518208
+```
+"""
+function cdf(d::Folded, x::Real) 
+    if !insupport(d,x)
+        (x ≤ minimum(d)) && return zero(x) # Is below the support
+        (x ≥ maximum(d)) && return one(x) # Is above the support
+    end
+    d.included.tp*cdf(d.included, x) + d.excluded.tp*ccdf(d.excluded, unfold_value(x,d))
+end
+
+
+# Quantiles will be calculated by the newton method:
+@quantile_newton(Folded)
+
+
+######################################################
+#         Statistics extended methods
+######################################################
+
+"""
+    mean(d::Folded)
+
+Gives the mean of a folded distribution. This will only work if `mean` is implemented for
+the truncated version of the parent distribution.
+
+Example:
+```julia
+mean(folded(Normal(1,2),0)) # -> 1.7911862296052243
+```
+
+Note that if the mean for the truncated distribution is not implemented, then this will give a Method 
+Error:
+```
+mean(folded(Beta(1,2)),0) # Errors out because mean(truncated(Beta(1,2),..)) is not implemented
+# >> MethodError: no method matching iterate(::Truncated{Beta{...}...})
+```
+"""
+mean(d::Folded) = mean(d.included)*d.included.tp - mean(d.excluded)*d.excluded.tp + 2*d.crease
```

Comment:
Can you add comments or expand this out? It's not immediately obvious to me why this is correct.